pkg/controllers/infra/cluster_controller.go

```diff
@@ -70,19 +74,37 @@ func (r *ClusterController) Reconcile(ctx context.Context, req ctrl.Request) (_
 	}
 
 	defer func() {
-		if err := patchHelper.Patch(ctx, infraCluster); err != nil {
+		patchOpts := []patch.Option{
+			patch.WithOwnedConditions{Conditions: []capiv1.ConditionType{
+				infrav1.InfraProviderReadyCondition,
+				infrav1.CNIReadyCondition,
+				capiv1.ReadyCondition,
+			}},
+		}
+
+		if reterr != nil {
+			infraCluster.Status.Phase = string(infrav1.ClusterPhaseFailed)
+		}
+
+		if err := patchHelper.Patch(ctx, infraCluster, patchOpts...); err != nil {
 			reterr = utilerrors.NewAggregate([]error{reterr, errors.Wrapf(err, "failed to patch infra Cluster %s", req.NamespacedName)})
 		}
 	}()
 
 	// Add finalizer if not exist to void the race condition.
 	if !controllerutil.ContainsFinalizer(infraCluster, clusterFinalizer) {
+		infraCluster.Status.Phase = string(capiv1.ClusterPhaseProvisioning)
```

Comment:
confusing when it has no finalizer set provisioning, it should be when its phase unset, set it?